samples/02.react-video/__tests__/test1.js

```diff
@@ -0,0 +1,259 @@
+      this.eventCount++;
+    }
+  }
+
+  addPage(page, prefix) {
+    page.on("console", (msg) => {
+      this.onLogEvent(msg.text());
+      if (this.verbose) {
+        console.log(`${prefix} ${msg.text()}`);
+      }
+    });
+  }
+}
+
+const delay = (timeout) => {
+  return new Promise((resolve) => {
+    setTimeout(resolve, timeout);
+  });
+};
+
+// utilities for working with browser tabs
+const openNewWindow = async (page, url) => {
+  page2 = await global.__BROWSER__.newPage();
+
+  await page2.goto(url);
+  await page.bringToFront();
+  return page2;
+};
+
+const closeAllPages = async () => {
+  const pages = await global.__BROWSER__.pages();
+  for (i = 0; i < pages.lengh; i++) {
+    pages[i].close();
+  }
+};
+
+let pageNumber = 0;
+const trackPages = (logsWatcher) => {
+  global.__BROWSER__.on("targetcreated", async (target) => {
+    pageNumber++;
+    console.log(target);
+    const newPage = await target.page();
+    logsWatcher.addPage(newPage, "page" + pageNumber);
+  });
+};
+
+// utilities to control video play back
+const playVideo = async (page) => {
+  await page.click("#player");
+  await delay(3000);
+};
+
+const seekVideo = async (page, seekTo) => {
+  const duration = await page.evaluate(() => {
+    return document.getElementById("player").duration;
+  });
+  console.log(`seekTo ${seekTo} duration ${duration}`);
+  const seekRatio = Math.min(1, seekTo / duration);
+
+  const element = await page.$("#slider-6");
+  const box = await element.boundingBox();
+  console.log(box);
+  // move mouse to slider start
+  await page.mouse.move(box.x, box.y, { steps: 5 });
+  await page.mouse.down();
+  // drag to seekTo location
+  const finalMouseX = box.x + box.width * seekRatio;
+  console.log(` move x ${finalMouseX}`);
+  await page.mouse.move(finalMouseX, box.y);
+  await page.mouse.up();
+  await delay(1000);
+};
+
+describe("/ (Home Page)", () => {
+  let page;
+  let logsWatcher;
+  let numberOfWindows = 5;
+  beforeAll(async () => {
+    logsWatcher = new LogsWatcher(true);
+    trackPages(logsWatcher);
+    page = await global.__BROWSER__.newPage();
+    await installMouseHelper(page);
+    await page.goto("http://localhost:3000");
+  });
+
+  afterAll(async () => {});
+
+  const verifyEvent = async (eventName, eventMatcher, performAction) => {
+    logsWatcher.resetTracking();
+    logsWatcher.startTracking(eventMatcher);
+    await performAction();
+    expect(logsWatcher.getEventCount()).toBe(numberOfWindows);
+    console.log(chalk.green(`Successfully verified ${eventName}`));
+  };
+
+  it(
+    "should play, seek, pause video for all clients",
+    async () => {
+      await delay(2000);
+      for (i = 1; i != numberOfWindows; i++) {
+        await openNewWindow(page, page.url());
+      }
+      await delay(1000);
+
+      await verifyEvent("play", isPlay, async () => {
+        await playVideo(page);
+      });
+
+      await delay(1000);
+
+      await verifyEvent(
+        "seek",
+        (log) => {
```

Comment:
are all tests based on emitted logs? is it possible to listen to WS in the tests to get more reliable tests?

Reply:
would like to understand how switching to websockets will make the tests more reliable.
we can listen on websockets but that would be low level and require more complex code to maintain.
also, we would like this to be something 3p devs can use and replicate. working websocket level data may not be appealing to 3p devs. logs provide high level event that we want to verify and will also be appealing to 3p.